### DISCUSSION TITLE
logging and rendering a table

### DISCUSSION THREAD

**Discussion by auphofBSF** [Textualize/rich#1799] [Q&A] [answered]
firstly @willmcgugan , this is a fantastic useful library thank you.
Now I am trying have a rich.table rendered in a log entry with the styles defined in the table and it is producing a result like this.
![image](https://user-images.githubusercontent.com/11743864/148124893-2dd7cf9b-e224-4e08-a060-897eece5a721.png)

I have got some partial success doing the following but the styling disappears
```
capture_text =  Text.from_ansi(capture.get())
log.debug(f"a table\n{capture_text}")
```
![image](https://user-images.githubusercontent.com/11743864/148125105-fce6373d-b4b4-4ef3-a14c-971eabb8fe87.png)

I am sure there is some subtle trick I am missing to make it look like this
![image](https://user-images.githubusercontent.com/11743864/148126301-ea7a2455-2e4a-45a4-bd33-b684efdb6e42.png)

My full experimental code is here
```python
import logging
from rich.logging import RichHandler

from rich.console import Console
from rich.table import Table
from rich.text import Text


logging.basicConfig(
    level=logging.DEBUG,
    handlers=[RichHandler(console = Console(width=150))] #width workaround for IPython default to 80 char
    # force=True
    )
log= logging.getLogger(__name__)



# %%
log.info("test") # This works great
log.error("[bold red blink]Server is shutting down![/]", extra={"markup": False})
log.error("[bold red blink]Server is shutting down![/]", extra={"markup": True})
# %%
print("-"*150)
# %%
def a_rich_table():
    table = Table(title="Star Wars Movies")

    table.add_column("Released", justify="right", style="cyan", no_wrap=True)
    table.add_column("Title",style="magenta")
    table.add_column("Box Office", justify="right", style="green")

    table.add_row("Dec 20, 2019", "Star Wars: The Rise of Skywalker", "$952,110,690")
    table.add_row("May 25, 2018", "Solo: A Star Wars Story", "$393,151,347")
    table.add_row("Dec 15, 2017", "Star Wars Ep. V111: The Last Jedi", "$1,332,539,889")
    table.add_row("Dec 16, 2016", "Rogue One: A Star Wars Story", "$1,332,439,889")
    return table


# %%
#THIS WORKS Table column styles rendered perfectly
console = Console(width=150)
console.print(a_rich_table())


# %%
# THIS FAILS - prints the object id
log.debug(f"a table\n{a_rich_table()}")

# %%
#use capture to generate a string  
console = Console(width=150)
with console.capture() as capture:
    console.print(a_rich_table())


# %%
#THIS FAILS Table rendered as raw text with control chars
log.debug(f"a table\n{capture.get()}")

# %%
#THIS WORKS however looses the Table column styles
capture_text =  Text.from_ansi(capture.get())
log.debug(f"a table\n{capture_text}")


```

**Comment by willmcgugan** [comment, marked as answer]
Rich may handle the rendering in RichHandler, but Python logging is not designed for anything other that simple strings.

Keep in mind that logging can be configured to go anywhere: file, to syslog, to mail, etc. most of which don't support ansi control codes for color, and may break any formatting that Rich does. Which is why it is best to stick to simple string messages.

If you are using logging purely to write to the console, consider using `console.log` which can display Rich objects without any clever capturing of the output.

**Reply by auphofBSF** [reply]
Figured that was the case, 
I note your recommendation of `console.log`  The workaround below enables a debug log entry to including a table without the table styling , and that alone is super useful. Is there any better alternative way to do this in Rich, 
Thank you once again.

```python

def log_table(rich_table):
  """Generate an ascii formatted presentation of a Rich table
  Eliminates any column styling
  """
  console = Console(width=150)
  with console.capture() as capture:
      console.print(rich_table())
  return Text.from_ansi(capture.get())

# where `my_debug_rich_table` is a rich table
log.debug(f"a table\n{log_table(my_debug_rich_table)}")

```

**Comment by yanzhiemq** [comment]
`Text.from_ansi(capture.get()).markup` will return a markup string with the original styling

**Comment by adcox** [comment]
I do most of my work in a terminal with output printed to the command line - being able to log renderables is really useful to me, and it seems pretty straightforward by making a few modifications to the `RichHandler` and building a custom `Formatter` object. The python logging API is pretty content to pass along objects as messages until the final step (formatting), which makes this possible.

<img width="591" height="134" alt="image" src="https://github.com/user-attachments/assets/2dcaa4e8-0198-42ff-8a2a-13ec3b6700d7" />

The example below creates two custom objects, the `RenderableHandler`, derived from `RichHandler`, and a `RenderableFormatter`. They must be used together, as in the "application code" section, to display the renderables. If you instead use a `logging.StreamHandler` (with no formatter) in place of the `RenderableHandler` (a pretty standard, default configuration), the output is less helpful but not erroneous:

<img width="292" height="34" alt="image" src="https://github.com/user-attachments/assets/7eb989fd-12f4-4cc3-88ab-a71e4d9261b8" />

<hr>

**Code**

```python
#!/usr/bin/env python3
import logging

from rich.abc import RichRenderable
from rich.logging import RichHandler

# ------------------------------------------------------------------------------
# Modified Logging Classes
# ------------------------------------------------------------------------------
# Modify the RichHandler to pass Renderable straight through without transforming
# them into Text
class RenderableHandler(RichHandler):

    def render_message(self, record, message):
        if isinstance(message, RichRenderable):
            return message
        else:
            return super().render_message(record, message)

# Modify the formatter to pass Renderable straight through without casting to str
class RenderableFormatter(logging.Formatter):

    def format(self, record):
        if isinstance(record.msg, RichRenderable):
            return record.msg
        else:
            return super().format(record)

# ------------------------------------------------------------------------------
# Library Code
# ------------------------------------------------------------------------------
logger = logging.getLogger("lib.module")
def doStuff():
    logger.info("Hello world")

    from rich.table import Table
    table = Table("Col 1", "Col 2", title="A Title")
    table.add_row("Val 1", "Val 2", style="green")
    table.add_row("Val 3", "Val 4", style="blue")
    logger.info(table)

# ------------------------------------------------------------------------------
# Application Code
# ------------------------------------------------------------------------------
if __name__ == "__main__":
    # Configure the logger to display renderables
    log = logging.getLogger("lib")
    handler = RenderableHandler(show_path=False)
    handler.setFormatter(RenderableFormatter())
    log.addHandler(handler)
    log.setLevel(logging.INFO)


    doStuff()
```
